dbms/src/Flash/Pipeline/Schedule/TaskQueues/ResourceControlQueue.cpp

```diff
@@ -106,28 +136,24 @@ bool ResourceControlQueue<NestedTaskQueueType>::take(TaskPtr & task)
                 is_finished,
                 group_info.task_queue->empty());
 
-            // When highest priority of resource group is less than zero, means RU of all resource groups are exhausted.
-            // Should not take any task from nested task queue for this situation.
-            if (ru_exhausted)
-                break;
-
-            if (group_info.task_queue->empty())
+            if unlikely (!error_resource_groups.empty())
             {
-                // Nested task queue is empty, continue and try next resource group.
-                size_t erase_num = resource_group_task_queues.erase(group_info.name);
-                RUNTIME_CHECK_MSG(
-                    erase_num == 1,
-                    "cannot erase corresponding TaskQueue for task of resource group {}, erase_num: {}",
-                    group_info.name,
-                    erase_num);
-                resource_group_infos.pop();
+                auto iter = error_resource_groups.find(group_info.name);
+                if likely (iter != error_resource_groups.end())
+                {
+                    mustTakeTask(group_info.task_queue, task);
+                    lock.unlock();
+                    task->onErrorOccurred(iter->second);
+                    lock.lock();
```

Comment:
how about:
```
{
    lock_guard lock;
    while (true)
    {
        ...
        if has error task: set exception_ptr;
            break;
    }
}

if unlikely (exception_ptr)
    task->onErrorOccurred(exception_ptr);
return true;
```

Reply:
```
auto exception_ptr;
SCOPE_EXIT({
        if unlikely (exception_ptr)
            task->onErrorOccurred(exception_ptr);
    });

origin code.
```
Maybe it would be simpler

Reply:
ok

Reply:
done